Improve table docs

timhoffm opened the pull request:
## PR Summary

Changes:

- Documented `Cell`.
- Rewrote the module docstring.

Do we need the author, copyright, license stuff in the module docstring? Or are the redundant, because this stuff is handled on the project-level?

<s>I could not make sense of the following two sentences in the module docstring. Therefore I removed that information. If anyone knows what these should mean, I'd be happy to add a more clear description.

> The grid need not be rectangular

> You can add additional cells outside this [`(max_row, max_col)`] range to have convenient
ways of positioning more interesting grids.
</s>

anntzer commented:
See https://github.com/matplotlib/matplotlib/commit/735f3083635f67da2fd1b5d305a2806583244917#diff-3ca3e0d4e31b414f1e771501e9bf69d3L3 for how I handled authorship issues.  I have no problem with leaving in who originally contributed the module, but they're obviously not the sole author anymore.

ImportanceOfBeingErnest commented:
I interprete the sentence 

> The grid need not be rectangular. You can add additional cells outside this [(max_row, max_col)] range to have convenient ways of positioning more interesting grids.

such that you are not restricted to the initial grid of the table. E.g. if you created a 11x2 table, you may still add cells outside the range of [0,1] columns and [0,10] rows. E.g.

![image](https://user-images.githubusercontent.com/23121882/45923120-78e31d00-bedd-11e8-9956-eb531a47cab5.png)

<details>
 <summary>Code</summary>

```
import matplotlib.pyplot as plt

fig, ax = plt.subplots()

c1 = ["pale", "dark"]
c2 = ["green", "turquoise"]
c3 = ["crimson", "bisque"]

a = [["".join((c1[i%2],c2[(j+i)%2])) for i in range(2)] for j in range(11)]

table = ax.table(cellColours=a, loc="center left", colWidths=[0.05]*2)
h = table.get_celld()[(0,0)].get_height()
for j in range(5):
    for i in range(1+j,10-j):
        table.add_cell(i,j+2, 0.05, h, facecolor=c3[(i+j)%2])
        table.add_cell(i,10-j, 0.05, h, facecolor=c3[(i+j)%2])
        table.add_cell(12-j,i+1, 0.05, h, facecolor=c3[(i-j-1)%2])

plt.show()
```
</details>

timhoffm commented:
@ImportanceOfBeingErnest Thanks, I've added that information to the `Table` class docstring.

timhoffm commented:
> See [735f308#diff-3ca3e0d4e31b414f1e771501e9bf69d3L3](https://github.com/matplotlib/matplotlib/commit/735f3083635f67da2fd1b5d305a2806583244917#diff-3ca3e0d4e31b414f1e771501e9bf69d3L3) for how I handled authorship issues. I have no problem with leaving in who originally contributed the module, but they're obviously not the sole author anymore.

@anntzer Thanks, I've now used that style and converted author and copyright there. I've removed the license, because it just says "matplotlib license", which is the license of the whole project anyway. Not sure if we have to keep the copyright notice.

anntzer commented:
No opinion there; @tacaswell?

anntzer commented:
Letting @tacaswell have a look at the copyright issue, if he cares.

timhoffm commented:
@tacaswell explicit question: What to do with copyright notices in docstrings:

https://github.com/matplotlib/matplotlib/pull/12218/files#diff-cd21097aa8cce79ca8a23b7f52c48f99

- Keep them in the docstring
- Move them from the docstring to a comment in the file, so that they do not show up in the docs but are still in the source code
- Remove them because they are not necessary in that context

tacaswell commented:
I do not think we can remove them.

I think the best path forward is to add "Original code by:" above it and "subsequent changes are copyright Matplotlib Developers" and to move it out of the docstring.

timhoffm commented:
Rebased.

anntzer commented on lib/matplotlib/table.py:
```diff
@@ -29,9 +29,38 @@
-    PAD = 0.1  # padding between text and rectangle
+
+    PAD = 0.1
+    """Padding between text and rectangle."""
```
Kind of prefer leaving this as inline comment (fwiw).
(As a side note, compare with the padding of Annotations, which is also hardcoded
```
                    pad = renderer.points_to_pixels(4)
```
)

timhoffm replied:
Since this is a public attribute, it [shows up in the docs](https://matplotlib.org/devdocs/api/table_api.html?highlight=table#matplotlib.table.Cell.PAD)

I very much prefer that every element show the docs has a description. This is achived [using the docstring](https://15585-1385122-gh.circle-artifacts.com/0/home/circleci/project/doc/build/html/api/table_api.html?highlight=table#matplotlib.table.Cell.PAD)

I would be fine not documenting `PAD`, but only if we also remove it from the docs (either by excluding explicitly (not sure if that's feasible with autodoc in this context), or preferably by renaming to `_PAD`.

anntzer replied:
Oh, I didn't realize sphinx would pick it up.  I guess that works, then...

anntzer commented on lib/matplotlib/table.py:
```diff
@@ -42,7 +71,7 @@ def __init__(self, xy, width, height,
 
         # Call base
-        Rectangle.__init__(self, xy, width=width, height=height,
+        Rectangle.__init__(self, xy, width=width, height=height, fill=fill,
```
I guess that's an intended new feature (respecting `fill` instead of dropping it)?.

timhoffm replied:
Yes. I assume that is was accidentally omitted. There's no point in having `fill` in `__init__` if it's silently dropped.

anntzer commented on lib/matplotlib/table.py:
```diff
@@ -124,18 +154,27 @@ def _set_text_position(self, renderer):
     def get_text_bounds(self, renderer):
-        """ Get text bounds in axes co-ordinates. """
+        """
+        Return the bounds of the text as *(x, y, width, height)* in table
```
"Return text bounds etc." fits in one line.

anntzer commented on lib/matplotlib/table.py:
```diff
@@ -209,6 +248,14 @@ class Table(Artist):
+    The table consists of a grid of cells, which are indexed by (row, column).
+
+    For a simple table, you'll have a full grid of cells with indices from
+    (0, 0) to (max_rows-1, max_cols-1), in which the cell (0, 0) is positioned
```
(n_rows-1, n_cols-1) or (max_row, max_col) (by definition max_foo is the largest value of foo...).

anntzer commented on lib/matplotlib/table.py:
```diff
@@ -29,9 +29,38 @@
 class Cell(Rectangle):
     """
-    A cell is a  `.Rectangle` with some associated text.
+    A cell is a  `.Rectangle` with some associated `.Text`.
```
extra space after "is a"

anntzer commented:
Sorry, didn't pay attention to the milestone.
